clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,330 @@
+            this.heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.resetTimer();
+            this.membershipManager.updateState(response.data());
+            return;
+        }
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+                                 final long currentTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        Errors error = Errors.forCode(response.data().errorCode());
+        maybeTransitionToFailureState(error);
+        String errorMessage = response.data().errorMessage();
+        switch(error) {
+            case NOT_COORDINATOR:
+            case COORDINATOR_NOT_AVAILABLE:
+                String message = String.format("Coordinator node %s is either not started or not valid. Retrying",
+                    coordinatorRequestManager.coordinator());
+                logInfo(errorMessage, response, currentTimeMs);
+                coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+                break;
+            case COORDINATOR_LOAD_IN_PROGRESS:
+                // retry
+                message = String.format("Heartbeat was not successful because the coordinator node %s is " +
+                    "loading."
+                    + "Retrying", coordinatorRequestManager.coordinator());
+                logInfo(errorMessage, response, currentTimeMs);
+                break;
+            case GROUP_AUTHORIZATION_FAILED:
+                GroupAuthorizationException exception =
+                    GroupAuthorizationException.forGroupId(membershipManager.groupId());
+                logger.error("GroupHeartbeatRequest failed due to group authorization failure: {}", exception.getMessage());
+                nonRetriableErrorHandler.handle(exception, errorMessage);
+                break;
+            case INVALID_REQUEST:
+                logger.error("GroupHeartbeatRequest failed due to invalid request error: {}", errorMessage);
+                nonRetriableErrorHandler.handle(Errors.INVALID_REQUEST.exception(), errorMessage);
+                break;
+            case GROUP_MAX_SIZE_REACHED:
+                logger.error("GroupHeartbeatRequest failed due to the max group size limit: {}", errorMessage);
+                nonRetriableErrorHandler.handle(Errors.GROUP_MAX_SIZE_REACHED.exception(), errorMessage);
+                break;
+            case UNSUPPORTED_ASSIGNOR:
+                logger.error("GroupHeartbeatRequest failed due to unsupported assignor {}: {}",
+                    membershipManager.assignorSelection(), errorMessage);
+                nonRetriableErrorHandler.handle(Errors.UNSUPPORTED_ASSIGNOR.exception(), errorMessage);
+                break;
```

Comment:
I also wonder if having a tailored error message for each error is needed. An alternative would be to group errors and have a generic message for the fatal ones for instance. What do you think?

Reply:
Yap that's a good idea to make the code easier to follow. We do want to provide some context such as group id or group instance id to GROUP_AUTHORIZATION_FAILED and UNRELEASED_INSTANCE_ID respectively. So I'm grouping the following 4 errors

```
case INVALID_REQUEST:
case GROUP_MAX_SIZE_REACHED:
case UNSUPPORTED_ASSIGNOR:
case UNSUPPORTED_VERSION:
```